### ISSUE TITLE
request.defaults() doesn't always preserve headers and cookie jar

### ISSUE THREAD

**Issue by RaceProUK** [request/request#1581]
```javascript
var baseRequest = request.defaults({
  headers: {x-token: 'my-token'}
})

var specialRequest = baseRequest.defaults({
  headers: {special: 'special value'}
})
```

According to the documentation, `specialRequest` should send both the `x-token` and `special` headers. However, this isn't always the case; sometimes, it only sends the `special` header. Similarly, the cookie jar, if one is being used, gets unassigned.

I can provide a more concrete example if required.

**Comment by simov** [comment]
@RaceProUK more thorough example would be great :+1:

**Comment by RaceProUK** [comment]
As requested; I've stripped out stuff that's irrelevant from both snippets.

This is where we're creating the initial `request` object:
```javascript
var jar = request.jar(),
    browser = request.defaults({
        rejectUnauthorized: false,
        jar: jar,
        headers: {
            'X-Requested-With': 'XMLHttpRequest',
            'User-Agent': 'UserAgent'
        }
    });
```

And this is where we're trying to add a new header:
```javascript
browser = browser.defaults({
        headers: {
            'X-CSRF-Token': csrf
        }
    });
```

By the documentation, `browser` should now be sending all three headers as well as the contents of the cookie jar; however, when checking with `request-debug`, after the second snippet had executed, only the `X-CSRF-Token` header was set, and there were no cookies.

We've been able to work around it by doing this:
```javascript
browser = browser.defaults({
        headers: {
            'X-Requested-With': 'XMLHttpRequest',
            'User-Agent': 'UserAgent',
            'X-CSRF-Token': csrf
        },
        jar: jar
    });
```
Which works, but obviously we'd rather not duplicate code if we can help it :wink:

The version of `request` I spotted this issue in is 2.55.0.

**Comment by simov** [comment]
Thanks, here is what I'm testing with:

```js
var jar = request.jar(),
    browser = request.defaults({
      rejectUnauthorized: false,
      jar: jar,
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
        'User-Agent': 'UserAgent'
      }
    });

    browser = browser.defaults({
      headers: {
        'X-CSRF-Token': 'csrf'
      }
    });

    browser('http://requestb.in/ougkjpou', function (err, res, body) {
      if (err) console.log(err)
      console.log(res.statusCode)
    })
```

Is that the correct [result](http://requestb.in/ougkjpou?inspect)? I'm looking at the headers.

**Comment by RaceProUK** [comment]
The headers are there alright; not sure why I'm seeing different results…

Maybe it's because we're using `browser.get()` for the request that gets the CSRF token, then `browser.post()` for the next request? A bit like this:
```javascript
var jar = request.jar(),
    browser = request.defaults({
      rejectUnauthorized: false,
      jar: jar,
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
        'User-Agent': 'UserAgent'
      }
    });
    browser.get(getUrl, callback);
//other stuff goes on; eventually getting to this bit
    browser = browser.defaults({
      headers: {
        'X-CSRF-Token': 'csrf'
      }
    });
    browser.post(postUrl, body, callback);
```

**Comment by simov** [comment]
Ok, I was able to reproduce, probably that's a bug, I'll take a look at it.

**Comment by simov** [comment]
Ok, this was fixed in #1540 but it's not published to NPM yet :)

**Comment by simov** [comment]
Closing @RaceProUK this fix will be released when 2.56 is published on NPM
